# Shark Attacks

In [1]:
from IPython.display import Image
from IPython.core.display import HTML 
Image(url= "https://i.ytimg.com/vi/Hrp3EjCYAac/maxresdefault.jpg", width=800, height=800)

### Imports
Librerías

In [2]:
import pandas as pd


Abrimos el archivo CSV fuente obtenido de la siguiente URL: 

https://www.kaggle.com/teajay/global-shark-attacks/version/1

Datos en crudo

In [3]:
# NOTA: He ido probando diferentes encoding ya que utf-8 no es aceptada en algunos de los registros. 

data = pd.read_csv('GSAF5.csv', encoding='ISO-8859-1')

### Análisis preliminar de los datos
Cabecera

In [4]:
data.head(10)

Case Number       Date  Year        Type        Country  \
0  2016.09.18.c  18-Sep-16  2016  Unprovoked            USA   
1  2016.09.18.b  18-Sep-16  2016  Unprovoked            USA   
2  2016.09.18.a  18-Sep-16  2016  Unprovoked            USA   
3    2016.09.17  17-Sep-16  2016  Unprovoked      AUSTRALIA   
4    2016.09.15  16-Sep-16  2016  Unprovoked      AUSTRALIA   
5  2016.09.15.R  15-Sep-16  2016        Boat      AUSTRALIA   
6    2016.09.11  11-Sep-16  2016  Unprovoked            USA   
7    2016.09.07  07-Sep-16  2016  Unprovoked            USA   
8    2016.09.06  06-Sep-16  2016  Unprovoked  NEW CALEDONIA   
9  2016.09.05.b  05-Sep-16  2016  Unprovoked            USA   

                Area                                         Location  \
0            Florida                 New Smyrna Beach, Volusia County   
1            Florida                 New Smyrna Beach, Volusia County   
2            Florida                 New Smyrna Beach, Volusia County   
3           Victoria                                 Thirteenth Beach   
4           Victoria                                      Bells Beach   
5  Western Australia                                          Bunbury   
6            Florida                    Ponte Vedra, St. Johns County   
7             Hawaii                                     Makaha, Oahu   
8     North Province                                           Koumac   
9     South Carolina  Kingston Plantation, Myrtle Beach, Horry County   

          Activity                    Name Sex      ...              Species   \
0          Surfing                    male    M     ...                   NaN   
1          Surfing          Chucky Luciano    M     ...                   NaN   
2          Surfing                    male    M     ...                   NaN   
3          Surfing         Rory Angiolella    M     ...                   NaN   
4          Surfing                    male    M     ...             2 m shark   
5          Fishing  Occupant: Ben Stratton  NaN     ...                   NaN   
6           Wading                    male    M     ...        3' to 4' shark   
7         Swimming                  female    F     ...      Tiger shark, 10?   
8     Kite surfing            David Jewell    M     ...                   NaN   
9  Boogie boarding          Rylie Williams    F     ...                   NaN   

        Investigator or Source                        pdf  \
0  Orlando Sentinel, 9/19/2016       2016.09.18.c-NSB.pdf   
1  Orlando Sentinel, 9/19/2016   2016.09.18.b-Luciano.pdf   
2  Orlando Sentinel, 9/19/2016       2016.09.18.a-NSB.pdf   
3           The Age, 9/18/2016  2016.09.17-Angiolella.pdf   
4           The Age, 9/16/2016  2016.09.16-BellsBeach.pdf   
5   West Australian, 9/15/2016      2016.09.15.R-boat.pdf   
6          News4Jax, 9/11/2016  2016.09.11-PonteVedra.pdf   
7    Hawaii News Now, 9/7/2016        2016.09.07-Oahu.pdf   
8       TVANouvelles, 9/6/2016      2016.09.06-Jewell.pdf   
9            C. Creswell, GSAF  2016.09.05.b-Williams.pdf   

                                        href formula  \
0  http://sharkattackfile.net/spreadsheets/pdf_di...   
1  http://sharkattackfile.net/spreadsheets/pdf_di...   
2  http://sharkattackfile.net/spreadsheets/pdf_di...   
3  http://sharkattackfile.net/spreadsheets/pdf_di...   
4  http://sharkattackfile.net/spreadsheets/pdf_di...   
5  http://sharkattackfile.net/spreadsheets/pdf_di...   
6  http://sharkattackfile.net/spreadsheets/pdf_di...   
7  http://sharkattackfile.net/spreadsheets/pdf_di...   
8  http://sharkattackfile.net/spreadsheets/pdf_di...   
9  http://sharkattackfile.net/spreadsheets/pdf_di...   

                                                href Case Number.1  \
0  http://sharkattackfile.net/spreadsheets/pdf_di...  2016.09.18.c   
1  http://sharkattackfile.net/spreadsheets/pdf_di...  2016.09.18.b   
2  http://sharkattackfile.net/spreadsheets/pdf_di...  2016.09.18.a   
3  http://sharkattackfile.net/spreadsheets/pd

Datos estadísticos

In [5]:
# De aquí no podemos sacar muchos valores numéricos

stats = data.describe().transpose()
stats

count         mean          std  min      25%     50%  \
Year            5992.0  1925.204606   286.473712  0.0  1942.00  1975.0   
original order  5992.0  2997.548899  1729.860210  2.0  1499.75  2997.5   

                    75%     max  
Year            2003.00  2016.0  
original order  4495.25  5993.0

Nombres de las columnas 

In [6]:
data.columns

Index(['Case Number', 'Date', 'Year', 'Type', 'Country', 'Area', 'Location',
       'Activity', 'Name', 'Sex ', 'Age', 'Injury', 'Fatal (Y/N)', 'Time',
       'Species ', 'Investigator or Source', 'pdf', 'href formula', 'href',
       'Case Number.1', 'Case Number.2', 'original order', 'Unnamed: 22',
       'Unnamed: 23'],
      dtype='object')

### Limpieza de datos

Hay columnas con ESPACIO detrás del texto: hay que renombrarlas

In [7]:
# NOTA: Fallo que se pasó por alto las primeras iteraciones. 

data = data.rename(columns={'Sex ':'Sex', 
                            'Species ':'Species',
                            'Case Number': 'Case_Number', 
                            'Fatal (Y/N)': 'Fatal', 
                            'Investigator or Source': 'Source', 
                            'href formula': 'href_formula', 
                            'Case Number.1': 'Case_1', 
                            'Case Number.2': 'Case_2', 
                            'original order': 'order', 
                            'Unnamed: 22': 'Unnamed_22', 
                            'Unnamed: 23': 'Unnamed_23'
                           })

Columnas con valores NaN

In [8]:
null_cols = data.isnull().sum()
null_cols[null_cols > 0]

Country           43
Area             402
Location         496
Activity         527
Name             200
Sex              567
Age             2681
Injury            27
Fatal             19
Time            3213
Species         2934
Source            15
href_formula       1
href               3
Unnamed_22      5991
Unnamed_23      5990
dtype: int64

Eliminamos las columnas ['Unnamed_22', 'Unnamed_23'] puesto que no traen datos junto con las demás que traigan datos NaN en más de la mitad de sus registros

In [9]:
drop_cols = list(null_cols[null_cols > len(data)/2].index)
data1 = data.drop(drop_cols, axis=1)
data1

Case_Number                                               Date  Year  \
0     2016.09.18.c                                          18-Sep-16  2016   
1     2016.09.18.b                                          18-Sep-16  2016   
2     2016.09.18.a                                          18-Sep-16  2016   
3       2016.09.17                                          17-Sep-16  2016   
4       2016.09.15                                          16-Sep-16  2016   
5     2016.09.15.R                                          15-Sep-16  2016   
6       2016.09.11                                          11-Sep-16  2016   
7       2016.09.07                                          07-Sep-16  2016   
8       2016.09.06                                          06-Sep-16  2016   
9     2016.09.05.b                                          05-Sep-16  2016   
10    2016.09.05.a                                          05-Sep-16  2016   
11      2016.09.04                                          04-Sep-16  2016   
12      2016.09.01                                          01-Sep-16  2016   
13    2016.08.29.b                                          29-Aug-16  2016   
14    2016.08.29.a                                          29-Aug-16  2016   
15      2016.08.27                                          27-Aug-16  2016   
16      2016.08.25                                          25-Aug-16  2016   
17      2016.08.07                                          07-Aug-16  2016   
18      2016.08.06                                          06-Aug-16  2016   
19      2016.08.04                                          04-Aug-16  2016   
20      2016.07.29                                          29-Jul-16  2016   
21    2016.07.28.R                                          28-Jul-16  2016   
22      2016.07.28                                          28-Jul-16  2016   
23      2016.07.27                                          27-Jul-16  2016   
24      2016.07.26                                          26-Jul-16  2016   
25      2016.07.24                                          24-Jul-16  2016   
26    2016.07.23.b                                          23-Jul-16  2016   
27    2016.07.23.a                                          23-Jul-16  2016   
28      2016.07.20                                          20-Jul-16  2016   
29      2016.07.17                                          17-Jul-16  2016   
...            ...                                                ...   ...   
5962       ND.0031                                       World War II     0   
5963       ND.0030                                        Before 1905     0   
5964       ND.0028                            A few years before 1938     0   
5965       ND.0027                                            No date     0   
5966       ND.0026                                        Early 1930s     0   
5967       ND.0025                                        Before 1927     0   
5968       ND.0024                                Between 1918 & 1939     0   
5969       ND.0023                                            No date     0   
5970       ND.0022                                            No date     0   
5971       ND.0021                                            No date     0   
5972       ND.0020                                         1920 -1923     0   
5973       ND.0019                                        Before 1921     0   
5974       ND.0018                                        Before 1911     0   
5975       ND.0017                                        Before 1921     0   
5976       ND.0016                                        Before 1921     0   
5977       ND.0015                                        Before 1917     0   
5978       ND.0014                        Before          17-Jul-1916     0   
5979       ND.0013  No date (3 days after preceding incident) & pr...     0   
5980       ND.0012                        Before         

Ademas de ['Unnamed_22', 'Unnamed_23'] se han eliminado las siguientes columnas: 

In [10]:
for e in ['Unnamed_22', 'Unnamed_23']: 
    drop_cols.remove(e)
drop_cols

['Time']

In [11]:
data1.columns

Index(['Case_Number', 'Date', 'Year', 'Type', 'Country', 'Area', 'Location',
       'Activity', 'Name', 'Sex', 'Age', 'Injury', 'Fatal', 'Species',
       'Source', 'pdf', 'href_formula', 'href', 'Case_1', 'Case_2', 'order'],
      dtype='object')

Vamos a quedarnos con las columnas que nos interesan nada más: 

In [12]:
data1 = data1[['Year', 'Type', 'Country', 'Area', 'Location',
       'Activity', 'Injury', 'Fatal', 'Species',
       'Source']]
data1

Year          Type             Country                       Area  \
0     2016    Unprovoked                 USA                    Florida   
1     2016    Unprovoked                 USA                    Florida   
2     2016    Unprovoked                 USA                    Florida   
3     2016    Unprovoked           AUSTRALIA                   Victoria   
4     2016    Unprovoked           AUSTRALIA                   Victoria   
5     2016          Boat           AUSTRALIA          Western Australia   
6     2016    Unprovoked                 USA                    Florida   
7     2016    Unprovoked                 USA                     Hawaii   
8     2016    Unprovoked       NEW CALEDONIA             North Province   
9     2016    Unprovoked                 USA             South Carolina   
10    2016    Unprovoked           AUSTRALIA          Western Australia   
11    2016    Unprovoked                 USA                    Florida   
12    2016    Unprovoked                 USA                 California   
13    2016    Unprovoked                 USA                    Florida   
14    2016    Unprovoked                 USA                    Florida   
15    2016    Unprovoked             REUNION                        NaN   
16    2016    Unprovoked                 USA                    Florida   
17    2016    Unprovoked             BAHAMAS      New Providence Island   
18    2016    Unprovoked                 USA                     Hawaii   
19    2016    Unprovoked                 USA                    Florida   
20    2016    Unprovoked               SPAIN          Alicante Province   
21    2016    Unprovoked               CHINA                  Hong Kong   
22    2016          Boat           AUSTRALIA          Western Australia   
23    2016      Provoked                 USA                    Florida   
24    2016    Unprovoked           AUSTRALIA            New South Wales   
25    2016    Unprovoked               JAPAN           Kochi Prefecture   
26    2016    Unprovoked           AUSTRALIA                   Tasmania   
27    2016    Unprovoked             BAHAMAS              Abaco Islands   
28    2016      Provoked           AUSTRALIA                 Queensland   
29    2016          Boat                 USA                    Alabama   
...    ...           ...                 ...                        ...   
5962     0  Sea Disaster    PAPUA NEW GUINEA            Madang Province   
5963     0    Unprovoked               BURMA                        NaN   
5964     0          Boat               ITALY               Adriatic Sea   
5965     0    Unprovoked              GREECE         Dodecanese Islands   
5966     0    Unprovoked              BELIZE                        NaN   
5967     0    Unprovoked           AUSTRALIA            New South Wales   
5968     0    Unprovoked             REUNION                Saint-Denis   
5969     0    Unprovoked        SOUTH AFRICA      Western Cape Province   
5970     0    Unprovoked           AUSTRALIA                        NaN   
5971     0    Unprovoked        SOUTH AFRICA              KwaZulu-Natal   
5972     0    Unprovoked           AUSTRALIA                 Queensland   
5973     0    Unprovoked                 USA                    Florida   
5974     0    Unprovoked             VIETNAM  Ba Ria-Vung Tau  Province   
5975     0    Unprovoked        SOUTH AFRICA              KwaZulu-Natal   
5976     0    Unprovoked        SOUTH AFRICA              KwaZulu-Natal   
5977     0    Unprovoked                FIJI               Moala Island   
5978     0    Unprovoked                 USA            North Carolina    
5979     0    Unprovoked        SOUTH AFRICA              KwaZulu-Natal   
5980     0    Unprovoked        SOUTH AFRICA              KwaZulu-Natal   
5981     0    Unprovoked               ASIA?                        NaN   
5982     0    Unprovoked                 USA                     Hawaii   
5983     0    Unprovoked       

In [13]:
data1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5992 entries, 0 to 5991
Data columns (total 10 columns):
Year        5992 non-null int64
Type        5992 non-null object
Country     5949 non-null object
Area        5590 non-null object
Location    5496 non-null object
Activity    5465 non-null object
Injury      5965 non-null object
Fatal       5973 non-null object
Species     3058 non-null object
Source      5977 non-null object
dtypes: int64(1), object(9)
memory usage: 468.2+ KB


Se puede apreciar que hay valores muy parecidos. Vamos a unificar actividades como por ejemplo todo lo que empiece por "Surf" será 'Surfing', 'Swimming' o 'Bathing' pasarán a 'Swiming', etc. (Más adelante)

In [14]:
activities = data1.Activity.value_counts()
activities.head(30)

Surfing            904
Swimming           819
Fishing            414
Spearfishing       321
Bathing            153
Wading             144
Diving             123
Standing            97
Snorkeling          77
Scuba diving        74
Body boarding       54
Body surfing        48
Swimming            47
Treading water      32
Pearl diving        32
Kayaking            28
Boogie boarding     28
Free diving         26
Fell overboard      21
Windsurfing         18
Boogie Boarding     16
Walking             15
Shark fishing       15
Floating            12
Surf-skiing         12
Canoeing            12
Rowing              12
Fishing             12
Surf skiing         12
Surf fishing        11
Name: Activity, dtype: int64

Hay columnas con valores inválidos todavía.

In [15]:
null_cols = data1.isnull().sum()
null_cols[null_cols > 0]

Country       43
Area         402
Location     496
Activity     527
Injury        27
Fatal         19
Species     2934
Source        15
dtype: int64

Eliminamos los registros con valores no válidos

In [16]:
data_na_clean = data1.dropna()
data_na_clean

Year          Type           Country                             Area  \
4     2016    Unprovoked         AUSTRALIA                         Victoria   
6     2016    Unprovoked               USA                          Florida   
7     2016    Unprovoked               USA                           Hawaii   
12    2016    Unprovoked               USA                       California   
13    2016    Unprovoked               USA                          Florida   
18    2016    Unprovoked               USA                           Hawaii   
19    2016    Unprovoked               USA                          Florida   
20    2016    Unprovoked             SPAIN                Alicante Province   
22    2016          Boat         AUSTRALIA                Western Australia   
23    2016      Provoked               USA                          Florida   
27    2016    Unprovoked           BAHAMAS                    Abaco Islands   
28    2016      Provoked         AUSTRALIA                       Queensland   
29    2016          Boat               USA                          Alabama   
34    2016    Unprovoked             SPAIN                   Canary Islands   
35    2016          Boat               USA                       California   
36    2016      Provoked               USA                    Massachusetts   
37    2016          Boat               USA                       California   
39    2016      Provoked         AUSTRALIA                       Queensland   
43    2016    Unprovoked      SOUTH AFRICA            Western Cape Province   
44    2016    Unprovoked               USA                   South Carolina   
46    2016    Unprovoked               USA                           Hawaii   
47    2016      Provoked         AUSTRALIA                Western Australia   
48    2016    Unprovoked               USA                            Texas   
49    2016    Unprovoked               USA                   North Carolina   
50    2016       Invalid               USA                   South Carolina   
52    2016    Unprovoked         AUSTRALIA                Western Australia   
53    2016    Unprovoked             EGYPT                             Suez   
56    2016    Unprovoked         AUSTRALIA                Western Australia   
58    2016    Unprovoked               USA                          Florida   
60    2016    Unprovoked               USA                          Florida   
...    ...           ...               ...                              ...   
5737  1853    Unprovoked               USA                   South Carolina   
5739  1853    Unprovoked               USA                   North Carolina   
5749  1852  Sea Disaster      SOUTH AFRICA            Western Cape Province   
5750  1852    Unprovoked       NEW ZEALAND                     North Island   
5754  1851    Unprovoked               USA                       California   
5786  1840    Unprovoked               USA                   South Carolina   
5788  1839    Unprovoked              CUBA  Matanzas Province (north coast)   
5798  1834    Unprovoked  FRENCH POLYNESIA                  Society Islands   
5826  1812    Unprovoked           ENGLAND                            Devon   
5850  1721    Unprovoked             ITALY                         Sardinia   
5853  1700    Unprovoked            FRANCE                     Côte d'Azur    
5881     0    Unprovoked               USA                           Hawaii   
5884     0    Unprovoked               USA                       California   
5902     0          Boat        MOZAMBIQUE               Inhambane Province   
5906     0    Unprovoked           BAHAMAS                   Andros Islands   
5912     0    Unprovoked             SUDAN                    Red Sea State   
5914     0      Provoked           VANUATU                 Malampa Province   
5916     0      Provoked               USA                          Florida   
5917     0       Invalid  PAPUA NEW GUINEA              

Se puede comprobar que las columnas con valores nan son 0 ahora

In [17]:
null_cols = data_na_clean.isnull().sum()
null_cols[null_cols > 0]

Series([], dtype: int64)

### Mejora de los datos

Una vez eliminadas las columnas cuyos datos son inválidos y los registros que contienen valores también inválidos para nosotros o que no aporten información pasamos a mejorar la calidad de los datos

Con el ejemplo de diferentes actividades

In [18]:
activities = data_na_clean.Activity.value_counts()
activities.head(20)

Surfing            543
Swimming           329
Fishing            227
Spearfishing       213
Wading              76
Scuba diving        53
Snorkeling          53
Diving              53
Standing            45
Body boarding       36
Body surfing        24
Kayaking            23
Bathing             23
Swimming            21
Treading water      19
Boogie boarding     16
Free diving         14
Shark fishing       12
Surf-skiing         11
Windsurfing         10
Name: Activity, dtype: int64

Se aprecia que **data_na_clean** todavía contiene registros que pueden considerarse duplicados como: Scuba diving <-> Free diving

In [19]:
# NOTA: este proceso se puede realizar con pandas de una forma más fácil. 

In [20]:
# Ejemplo de diccionario para traducir actividades

# NOTA: Visitar
# http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
CONTAINS_ALIAS_ACTIVITY = {
    'Surfing': ['Surf', 'surf', 'board', 'Board'], 
    'Swimming': ['Swim', 'swim', 'snor', 'Snor', 'float', 'Float', 'Bath', 'bath'], 
    'Fishing': ['Fish', 'fish'], 
    'Diving': ['Divi', 'divi']
}
for key, value in CONTAINS_ALIAS_ACTIVITY.items():
    for act in value: 
        data_na_clean.loc[data_na_clean['Activity'].str.contains(act, regex=False), 'Activity'] = key

/home/alberto/.local/lib/python3.6/site-packages/pandas/core/indexing.py:543: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  self.obj[item] = s


In [21]:
activities = data_na_clean.Activity.value_counts()
activities.head(20)

Surfing                  844
Fishing                  662
Swimming                 524
Diving                   243
Wading                    76
Standing                  45
Kayaking                  23
Treading water            19
Walking                    8
Rowing                     7
Playing                    5
Paddleskiing               5
Boating                    5
Boat                       5
Canoeing                   4
Jumping                    4
Clamming                   3
Sculling                   3
Jumped into the water      3
Crabbing                   3
Name: Activity, dtype: int64

Se puede aprceiar que se han ido agrupando los Surfing. A medida que se incremente el diccionario CONTAINS_ALIAS_ACTIVITY irá mejorando la solución


Con el ejemplo de diferentes tipos de ataque podemos hacer algo similar a lo realizado con actividades. 

In [22]:
types = data_na_clean.Type.value_counts()
types

Unprovoked      1981
Provoked         358
Invalid          172
Boat             135
Boating           47
Sea Disaster      19
Name: Type, dtype: int64

In [23]:
CONTAINS_ALIAS_TYPE = {
    'Boating': ['Boat', 'boat']
}

# mejorable 
for key, value in CONTAINS_ALIAS_TYPE.items():
    for ty in value: 
        data_na_clean.loc[data_na_clean['Type'].str.contains(ty, regex=False), 'Type'] = key

In [24]:
types = data_na_clean.Type.value_counts()
types

Unprovoked      1981
Provoked         358
Boating          182
Invalid          172
Sea Disaster      19
Name: Type, dtype: int64

Podemos apreciar que Boating ahora agrupa Boat y Boating

### Conclusiones
A continuación obtendremos algunas conclusiones de nuestros datos y realizaremos operaciones que hemos estudiado en el Bootcamp

In [25]:
for c in data_na_clean.columns: 
    print(c, len(data_na_clean[c].value_counts()))

Year 154
Type 5
Country 99
Area 366
Location 2152
Activity 226
Injury 1919
Fatal 4
Species 1405
Source 2264


Se puede ver que de Species no se podrá obtener mucha información ya que no sigue ningún criterio. No hay catalogadas 1405 especies de tiburones así que habría que hacer un diccionario que tradujera los datos si queremos operar con esta columna

Datos estadísticos

In [26]:
data_na_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2712 entries, 4 to 5986
Data columns (total 10 columns):
Year        2712 non-null int64
Type        2712 non-null object
Country     2712 non-null object
Area        2712 non-null object
Location    2712 non-null object
Activity    2712 non-null object
Injury      2712 non-null object
Fatal       2712 non-null object
Species     2712 non-null object
Source      2712 non-null object
dtypes: int64(1), object(9)
memory usage: 233.1+ KB


#### 1. El lugar más letal para bañarse en cada uno de los 7 paises con más mortalidad por ataque de tiburón

Paises donde más ataques mortales hay ordenados de más a menos peligroso. 

In [27]:
df_paises = data_na_clean[data_na_clean['Fatal'] == 'Y']
ser_paises = df_paises['Country'].value_counts().keys()
ser_paises

Index(['AUSTRALIA', 'USA', 'SOUTH AFRICA', 'PAPUA NEW GUINEA', 'NEW ZEALAND',
       'REUNION', 'MEXICO', 'CROATIA', 'NEW CALEDONIA', 'ITALY', 'PANAMA',
       'NEW BRITAIN', 'GREECE', 'BAHAMAS', 'JAMAICA', 'FIJI', 'JAPAN',
       'HONG KONG', 'EGYPT', 'OKINAWA', 'BRAZIL', 'MOZAMBIQUE', 'SPAIN',
       'MONTENEGRO', 'VANUATU', 'MALTA', 'SEYCHELLES', 'AMERICAN SAMOA',
       'INDONESIA', 'CUBA', 'DOMINICAN REPUBLIC', 'SOUTH KOREA', 'INDIA',
       'CHILE', 'IRAQ', 'Sierra Leone', 'CHINA', 'PHILIPPINES',
       'FRENCH POLYNESIA', 'SAUDI ARABIA', 'SENEGAL', 'EL SALVADOR', 'FRANCE',
       'YEMEN', 'TONGA', 'SAN DOMINGO', ' TONGA', 'NICARAGUA', 'PORTUGAL',
       'Seychelles', 'PACIFIC OCEAN', 'NORTH PACIFIC OCEAN', 'HAITI',
       'WESTERN SAMOA', 'NEW GUINEA', 'SOLOMON ISLANDS', 'MARSHALL ISLANDS',
       'CYPRUS'],
      dtype='object')

Los 7 paises que han registrado más muertes por ataque de tiburón

In [28]:
l_7_paises_letales = list(ser_paises[0:7])
l_7_paises_letales

['AUSTRALIA',
 'USA',
 'SOUTH AFRICA',
 'PAPUA NEW GUINEA',
 'NEW ZEALAND',
 'REUNION',
 'MEXICO']

¿Dónde no ir a bañarse si vas a estos paises?

In [29]:
res = []
for pais in l_7_paises_letales: 
    res.append((pais, df_paises[df_paises['Country'] == pais]['Area'].value_counts().keys().max() ))
res

[('AUSTRALIA', 'Western Australia'),
 ('USA', 'Virginia'),
 ('SOUTH AFRICA', 'Western Province'),
 ('PAPUA NEW GUINEA', 'West New Britain Province'),
 ('NEW ZEALAND', 'South Island'),
 ('REUNION', 'd\x92Étang-Salé'),
 ('MEXICO', 'Veracruz')]

#### 2. Agrupamos los paises por la cantidad ataques sufridos en el histórico

In [30]:
df_paises_bin = data_na_clean['Country'].value_counts()
df_paises_bin.head(20)

USA                 1140
AUSTRALIA            643
SOUTH AFRICA         350
NEW ZEALAND           65
BAHAMAS               44
ITALY                 35
PAPUA NEW GUINEA      35
BRAZIL                26
MEXICO                22
SPAIN                 20
REUNION               19
NEW CALEDONIA         16
MOZAMBIQUE            16
EGYPT                 15
CROATIA               12
FRENCH POLYNESIA      12
CUBA                  11
FIJI                  10
CANADA                 8
PANAMA                 8
Name: Country, dtype: int64

In [31]:
# NOTA: leer documentación es importante para entender los errores

# https://pandas.pydata.org/pandas-docs/stable/generated/pandas.qcut.html
"""
duplicates : {default ‘raise’, ‘drop’}, optional
If bin edges are not unique, raise ValueError or drop non-uniques.
New in version 0.20.0.
"""

mpg_labels = ['Very Low', 'Low', 'Moderate', 'High', 'Very High']
bins = pd.qcut(df_paises_bin,6, labels=mpg_labels, duplicates='drop')
bins

USA                               Very High
AUSTRALIA                         Very High
SOUTH AFRICA                      Very High
NEW ZEALAND                       Very High
BAHAMAS                           Very High
ITALY                             Very High
PAPUA NEW GUINEA                  Very High
BRAZIL                            Very High
MEXICO                            Very High
SPAIN                             Very High
REUNION                           Very High
NEW CALEDONIA                     Very High
MOZAMBIQUE                        Very High
EGYPT                             Very High
CROATIA                           Very High
FRENCH POLYNESIA                  Very High
CUBA                              Very High
FIJI                                   High
CANADA                                 High
PANAMA                                 High
ENGLAND                                High
MARSHALL ISLANDS                       High
SCOTLAND                        

Hemos seleccionado qcut porque nuestra distribución tiene muchos valores atípicos (outliers): USA y Australia muchos ataque pero la gran mayoría del resto muy pocos

#### 3. Mejoras futuras

3.1. ranking de paises más tontos: ranking de paises donde los ataques son provocados

3.2. Realizar un diccionario de tiburones: white, whaler, bull, tiger